include/alpaka/queue/sycl/QueueGenericSyclBase.hpp

```diff
@@ -113,39 +85,29 @@ namespace alpaka::detail
 
         auto get_last_event() const -> sycl::event
         {
-            std::shared_lock<std::shared_mutex> lock{m_mutex};
             return m_last_event;
         }
 
         template<bool TBlocking, typename TTask>
         auto enqueue(TTask const& task) -> void
         {
             {
-                std::lock_guard<std::shared_mutex> lock{m_mutex};
-
-                clean_dependencies();
-
                 // Execute task
                 if constexpr(is_sycl_task<TTask> && !is_sycl_kernel<TTask>) // Copy / Fill
                 {
-                    m_last_event = task(m_queue, m_dependencies); // Will call queue.{copy, fill} internally
+                    m_last_event = task(m_queue); // Will call queue.{copy, fill} internally
                 }
                 else
                 {
                     m_last_event = m_queue.submit(
```

Comment:
This looks dangerously like a race condition.

Reply:
Do you mean between the call to `enqueue()` and a possible call to `get_last_event()` or `empty()` ?

Reply:
I'm worried about the assignment here. What happens if two threads enqueue work concurrently? Do we have a guarantee that `m_last_event` will actually refer to the last event?

Reply:
Ah, I see.

You are right, if multiple threads enqueue a task to the same queue, there is no guarantee about which of the two events would be stored in `m_last_event`.

Possible solutions:
1. add back the lock, and use it only in `enqueue()`, `empty()`, `get_last_event()` (and maybe `wait()`? probably not)
2. remove the `m_last_event` member and
  - implement `get_last_event()` with a call to `m_queue.ext_oneapi_submit_barrier()`
  - implement `empty()` with a call to `m_queue.ext_oneapi_empty()`
Unfortunately according to the [extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_queue_empty.asciidoc) itself, `ext_oneapi_empty()` is not supported by the OpenCL back0end :-(

Reply:
From the extension link:

> Currently support for OpenCL backend is limited, API introduced by this extension can be called only for **in-order queues** which doesn’t have discard_events property. Exception is thrown if new API is called on other type of queue. OpenCL currently doesn’t have an API to get queue status.

We are using in-order SYCL queues internally, so I think we should be fine, right?

Reply:
Good point.

I guess we can try and see if it works :-)